### ISSUE TITLE
ODistributedConfigurationChangedException when using Java API insert in two node scenario

### ISSUE THREAD

**Issue by mxns** [orientechnologies/orientdb#6779]
## OrientDB Version, operating system, or hardware.
- v2.2.7
## Operating System
- [ ] Linux
- [ ] MacOSX
- [ ] Windows
- [ ] Other Unix
- [x] Other: Docker version 1.11.1, build 5604cbe on OS X El Capitan
## Expected behavior and actual behavior

Expected: Running a two node scenario with tcp-ip discovery, when using Java document API for inserting document on one of the nodes, the OrientDB server should select a cluster that belongs to that node, and the insert should succeed

Actual: in some cases, the insert fails with the stack trace:

```
2016-10-04 14:04:48:076 INFO  Local node 'orientdb-lab-node1' is not the master for cluster 'event_3' (it is 'orientdb-lab-node2'). Reloading distributed configuration for database 'test-db' [ODistributedStorage]com.orientechnologies.orient.server.distributed.ODistributedConfigurationChangedException: Local node 'orientdb-lab-node1' is not the master for cluster 'event_3' (it is 'orientdb-lab-node2')
        DB name="test-db"
        DB name="test-db"
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at com.orientechnologies.orient.client.binary.OChannelBinaryAsynchClient.throwSerializedException(OChannelBinaryAsynchClient.java:428)
        at com.orientechnologies.orient.client.binary.OChannelBinaryAsynchClient.handleStatus(OChannelBinaryAsynchClient.java:379)
        at com.orientechnologies.orient.client.binary.OChannelBinaryAsynchClient.beginResponse(OChannelBinaryAsynchClient.java:261)
        at com.orientechnologies.orient.client.binary.OChannelBinaryAsynchClient.beginResponse(OChannelBinaryAsynchClient.java:153)
        at com.orientechnologies.orient.client.remote.OStorageRemote.beginResponse(OStorageRemote.java:2133)
        at com.orientechnologies.orient.client.remote.OStorageRemote$6.execute(OStorageRemote.java:543)
        at com.orientechnologies.orient.client.remote.OStorageRemote$6.execute(OStorageRemote.java:537)
        at com.orientechnologies.orient.client.remote.OStorageRemote$1.execute(OStorageRemote.java:166)
        at com.orientechnologies.orient.client.remote.OStorageRemote.baseNetworkOperation(OStorageRemote.java:230)
        at com.orientechnologies.orient.client.remote.OStorageRemote.asyncNetworkOperation(OStorageRemote.java:158)
        at com.orientechnologies.orient.client.remote.OStorageRemote.createRecord(OStorageRemote.java:523)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:2075)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.saveNew(OTransactionNoTx.java:246)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.saveRecord(OTransactionNoTx.java:179)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:2598)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:103)
        at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1800)
        at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1791)
        at lab.orientdb.OrientDbClient.insert(OrientDbClient.java:19)
        at lab.orientdb.Main.main(Main.java:25)
```
## Steps to reproduce the problem

The issue can be reproduced by running `docker-compose` setup in this project: 

https://github.com/mxns/orientdb-lab

Running this project as described in the README will start two OrientDB nodes in distributed mode, and start doing inserts on both nodes. The error appears as soon as start-up is complete and the program starts doing inserts, and the stack trace is printed to stdout.

Either build the images, or just run the `docker-compose.yaml` file, in which case the images will download from my repo at DockerHub.

In essence, what the (quite simple) test program does is the following: 
- Start two OrientDB nodes in distributed mode, node discovery by tcp-ip 
- Use following Java code to insert document:

```
    OPartitionedDatabasePool pool = new OPartitionedDatabasePool(dbUrl, user, password);
    ...
    try (ODatabaseDocumentTx ignored = pool.acquire()) {
        ODocument document = new ODocument(className);
        document.fromJSON(recordData.encode());
        document.field("date", timestamp, OType.DATETIME);
        document.save();
    }
```

Could perhaps be related to #6765, similar stack trace

**EDIT**: this is not a clear cut case. The error is quite consistent on the system that I am running. However, running on a newer version of Docker does not seem to have the same result. It might be an artifact of this specific (and pretty useless, except for testing) setup.

**EDIT2**: Doesn't seem to have to do with Docker version after all, but with the number of cores that the host vm is assigned. 1 core -> no problem, 4 cores -> problem.

**Comment by mxns** [comment]
Hi,

some further info: this behaviour is reproducible on physical machines, so it seems it is not just a docker thing. 

The company I work for is pretty invested in OrientDB, and we have a release coming up which will be the first where we have to support ODB in distributed mode. If you can confirm this as a bug, or otherwise, so we can make plans going forward, that would be awesome. Any workaround suggestion also welcome.

**Comment by lvca** [comment]
Hi @mxns, the problem could rely on the data in `recordData` variable. Does it contain also the class name? Can you try to remove it?

**Comment by mxns** [comment]
Hi @lvca thanks for looking into this. In the ´recordData´ is, in my test, just an empty json string, like this: ´{}´. I also tried removing it altogether, with the same result. The classname appears only in the `Document` constructor.

**Comment by lvca** [comment]
I cannot reproduce it with last 2.2.x. Could you please try with last snapshot (https://oss.sonatype.org/content/repositories/snapshots/com/orientechnologies/orientdb-community/2.2.12-SNAPSHOT/orientdb-community-2.2.12-20161012.150151-25.zip)?

**Comment by mxns** [comment]
I tested it with 2.2.11 with the same result. Testing with the zip file is a little bit more setup work to do, but sure I can try that too if there is any reason to think it might have changed since 2.2.11. 

Sorry for being a nuisance, but can you confirm that you have multiple clusters  in your test? My database looks like this: 

```
orientdb {db=test-db}> list clusters


CLUSTERS (collections)
+----+-----------+----+---------+-----------------+-----+------------------+--------------------+--------------------+
|#   |NAME       |  ID|CLASS    |CONFLICT-STRATEGY|COUNT|   OWNER_SERVER   |   OTHER_SERVERS    |AUTO_DEPLOY_NEW_NODE|
+----+-----------+----+---------+-----------------+-----+------------------+--------------------+--------------------+
|5   |event      |  17|event    |                 |   15|orientdb-lab-node2|[orientdb-lab-node1]|        true        |
|6   |event_1    |  18|event    |                 |   14|orientdb-lab-node1|[orientdb-lab-node2]|        true        |
|7   |event_2    |  19|event    |                 |   14|orientdb-lab-node1|[orientdb-lab-node2]|        true        |
|8   |event_3    |  20|event    |                 |   13|orientdb-lab-node2|[orientdb-lab-node1]|        true        
```

```
CONFIGURED SERVERS
+----+------------------+------+-----------+-------------------+------------------+------------------+----------------+-----------------+---------+
|#   |Name              |Status|Connections|StartedOn          |Binary            |HTTP              |UsedMemory      |FreeMemory       |MaxMemory|
+----+------------------+------+-----------+-------------------+------------------+------------------+----------------+-----------------+---------+
|0   |orientdb-lab-node2|ONLINE|4          |2016-10-17 10:55:52|172.16.238.12:2424|172.16.238.12:2480|98.09MB (11.17%)|41.41MB (4.71%)  |878.50MB |
|1   |orientdb-lab-node1|ONLINE|4          |2016-10-17 10:55:52|172.16.238.11:2424|172.16.238.11:2480|77.36MB (8.81%) |107.64MB (12.25%)|878.50MB |
+----+------------------+------+-----------+-------------------+------------------+------------------+----------------+-----------------+---------+
```

and the error message implies that node1 is trying to insert at cluster event_3, which it does not own:
```
com.orientechnologies.orient.server.distributed.ODistributedConfigurationChangedException: Local node 'orientdb-lab-node1' is not the owner for cluster 'event_3' (it is 'orientdb-lab-node2')
```

**Comment by mxns** [comment]
Resolved in 2.2.12, nice job @lvca!
